File: Tree/144BinaryTreePreorderTraversal.cpp
/**
 Pre-order Traversal
 Pre-order traversal is to visit the root first. Then traverse the left subtree. Finally, traverse the right subtree.
 
 Solution:
 Create a vector to store each node accessed. Use a helper function to traverse nodes recursively. Return NULL when current node is NULL.
 Step 1: Store the current node's value immediately.
 Step 2: Go to left node of current first.
 Step 3: Go to right node of current next.
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> rev;
        helper(root, rev);
        return rev;
    }
    void helper(TreeNode* root, vector<int> &rev){
        if(root==NULL) return;
        else{
            rev.push_back(root->val);
            helper(root->left, rev);
            helper(root->right, rev);
        }
    }
};

File: Tree/144binaryTreePreorderTraversal.py
#Key: binaryTree
#algorithm: Recursion

#The node must remember where it just arrive. so ans.append first, and then go to its left and right
#The recursion must have a base case, which is when it meet an end node, that's the reason if and else 

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ans = []
        def check(root):
            if not root:
                return
            else:
                ans.append(root.val)
                check(root.left)
                check(root.right)
        check(root)
        return ans